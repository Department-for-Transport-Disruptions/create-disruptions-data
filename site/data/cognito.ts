import {
    CognitoIdentityProviderClient,
    AdminInitiateAuthCommand,
    AdminInitiateAuthCommandInput,
    AdminInitiateAuthCommandOutput,
    AdminRespondToAuthChallengeCommand,
    AdminRespondToAuthChallengeCommandInput,
    AdminUserGlobalSignOutCommand,
    AdminUserGlobalSignOutCommandInput,
    AdminSetUserPasswordCommand,
    AdminSetUserPasswordCommandInput,
    AdminAddUserToGroupCommand,
    ListGroupsCommand,
    ListUsersInGroupCommand,
    UserType,
    AdminDeleteUserCommand,
    AdminGetUserCommand,
    AdminGetUserCommandInput,
    AdminDeleteUserCommandInput,
    AdminCreateUserCommand,
    ListUsersCommand,
    ConfirmForgotPasswordCommand,
    ConfirmForgotPasswordCommandInput,
    ForgotPasswordCommandInput,
    ForgotPasswordCommand,
} from "@aws-sdk/client-cognito-identity-provider";

import { createHmac } from "crypto";
import { AddUserSchema } from "../schemas/add-user.schema";
import { userManagementSchema } from "../schemas/user-management.schema";
import { notEmpty } from "../utils";
import logger from "../utils/logger";

const {
    COGNITO_CLIENT_ID: cognitoClientId,
    COGNITO_CLIENT_SECRET: cognitoClientSecret,
    COGNITO_USER_POOL_ID: userPoolId,
} = process.env;

if (!cognitoClientSecret || !cognitoClientId || !userPoolId) {
    throw new Error("Cognito env vars not set");
}

const cognito = new CognitoIdentityProviderClient({
    region: "eu-west-2",
});

export const deleteUser = (username: string) => {
    try {
        const params: AdminDeleteUserCommandInput = {
            UserPoolId: userPoolId,
            Username: username,
        };
        return cognito.send(new AdminDeleteUserCommand(params));
    } catch (error) {
        if (error instanceof Error) {
            throw new Error(`Failed to delete user: ${error.stack || ""}`);
        }

        throw error;
    }
};

export const getUserDetails = (username: string) => {
    try {
        const params: AdminGetUserCommandInput = {
            UserPoolId: userPoolId,
            Username: username,
        };
        return cognito.send(new AdminGetUserCommand(params));
    } catch (error) {
        if (error instanceof Error) {
            throw new Error(`Failed to get user details by username: ${error.stack || ""}`);
        }

        throw error;
    }
};

const calculateSecretHash = (username: string): string =>
    createHmac("SHA256", cognitoClientSecret)
        .update(username + cognitoClientId)
        .digest("base64");

export const initiateAuth = async (username: string, password: string): Promise<AdminInitiateAuthCommandOutput> => {
    logger.info("", {
        context: "data.cognito",
        message: "initiating auth",
    });

    try {
        const params: AdminInitiateAuthCommandInput = {
            AuthFlow: "ADMIN_USER_PASSWORD_AUTH",
            ClientId: cognitoClientId,
            UserPoolId: userPoolId,
            AuthParameters: {
                USERNAME: username,
                PASSWORD: password,
                SECRET_HASH: calculateSecretHash(username),
            },
        };

        return cognito.send(new AdminInitiateAuthCommand(params));
    } catch (error) {
        if (error instanceof Error) {
            throw new Error(`Failed to authenticate user: ${error.stack || ""}`);
        }

        throw error;
    }
};

export const respondToNewPasswordChallenge = async (
    username: string,
    password: string,
    session: string,
): Promise<void> => {
    logger.info("", {
        context: "data.cognito",
        message: "new password challenge initiated",
    });

    try {
        const params: AdminRespondToAuthChallengeCommandInput = {
            ChallengeName: "NEW_PASSWORD_REQUIRED",
            ClientId: cognitoClientId,
            UserPoolId: userPoolId,
            ChallengeResponses: {
                USERNAME: username,
                NEW_PASSWORD: password,
                SECRET_HASH: calculateSecretHash(username),
            },
            Session: session,
        };

        await cognito.send(new AdminRespondToAuthChallengeCommand(params));
    } catch (error) {
        if (error instanceof Error) {
            throw new Error(`Failed to respond to password challenge: ${error.stack || ""}`);
        }

        throw error;
    }
};

export const globalSignOut = async (username: string): Promise<void> => {
    logger.info("", {
        context: "data.cognito",
        message: "performing global sign out",
    });

    const params: AdminUserGlobalSignOutCommandInput = {
        Username: username,
        UserPoolId: userPoolId,
    };

    try {
        await cognito.send(new AdminUserGlobalSignOutCommand(params));
    } catch (error) {
        if (error instanceof Error) {
            throw new Error(`Failed to perform global sign out: ${error.stack || ""}`);
        }

        throw error;
    }
};

export const updateUserPassword = async (newPassword: string, username: string): Promise<void> => {
    logger.info("", {
        context: "data.cognito",
        message: "updating user password",
    });

    const params: AdminSetUserPasswordCommandInput = {
        Password: newPassword,
        Permanent: true,
        Username: username,
        UserPoolId: userPoolId,
    };

    try {
        await cognito.send(new AdminSetUserPasswordCommand(params));
    } catch (error) {
        if (error instanceof Error) {
            throw new Error(`Failed to update user password: ${error.stack || ""}`);
        }

        throw error;
    }
};

export const getAllUsersInGroup = async (groupName: string, nextToken?: string): Promise<UserType[]> => {
    const users = await cognito.send(
        new ListUsersInGroupCommand({ UserPoolId: userPoolId, GroupName: groupName, Limit: 60, NextToken: nextToken }),
    );

    if (!users.Users) {
        return [];
    }

    if (users.NextToken) {
        return [...users.Users, ...(await getAllUsersInGroup(groupName, users.NextToken))];
    } else {
        return users.Users;
    }
};

export const listUsersWithGroups = async () => {
    logger.info("", {
        context: "data.cognito",
        message: "Listing cognito users",
    });

    try {
        const groupList = await cognito.send(new ListGroupsCommand({ UserPoolId: userPoolId }));

        const userList = await Promise.all(
            groupList.Groups?.map(async (group) => {
                if (!group.GroupName) {
                    return null;
                }

                const groupUsers = await getAllUsersInGroup(group.GroupName);

                return groupUsers.map((user) => ({ ...user, GroupName: group.GroupName }));
            }) ?? [],
        );

        const parsedUsers = userManagementSchema.safeParse(userList.filter(notEmpty).flat());

        if (!parsedUsers.success) {
            return [];
        }

        return parsedUsers.data;
    } catch (error) {
        if (error instanceof Error) {
            throw new Error(`Failed to list cognito users: ${error.stack || ""}`);
        }

        throw error;
    }
};

export const deleteUsersByAttribute = async (attributeName: string, attributeValue: string) => {
    logger.info("", {
        context: "data.cognito",
        message: "Listing cognito users based on custom attributes and deleting the users that match the value",
    });

    try {
        const userList = await cognito.send(
            new ListUsersCommand({ UserPoolId: userPoolId, AttributesToGet: [attributeName] }),
        );

        const deletePromises = userList.Users?.filter((user) =>
            user.Attributes?.find((value) => value.Value === attributeValue),
        ).map((user) => {
            return cognito.send(
                new AdminDeleteUserCommand({
                    UserPoolId: userPoolId,
                    Username: user.Username,
                }),
            );
        });

        await Promise.all(deletePromises ?? []);
    } catch (error) {
        if (error instanceof Error) {
            throw new Error(
                `Failed to list cognito users based on attribute ${attributeName} with value ${attributeValue}: ${
                    error.stack || ""
                }`,
            );
        }

        throw error;
    }
};

export const createUser = async (userData: AddUserSchema) => {
    logger.info("", {
        context: "data.cognito",
        message: "Adding a new user",
    });

    const nocCodes = userData.operatorNocInfo?.map((operator) => operator.nocCode) ?? [];

    const userAttributes = [
        {
            Name: "custom:orgId",
            Value: userData.orgId,
        },
        {
            Name: "given_name",
            Value: userData.givenName,
        },
        {
            Name: "family_name",
            Value: userData.familyName,
        },
        {
            Name: "email_verified",
            Value: "true",
        },
        {
            Name: "email",
            Value: userData.email,
        },
    ];

    const createUserCommand = {
        Username: userData.email,
        UserPoolId: userPoolId,
        TemporaryPassword: Array.from(Array(20), () => Math.floor(Math.random() * 36).toString(36)).join(""),
        UserAttributes: userAttributes,
    };

    const createOperatorUserCommand = {
        ...createUserCommand,
        UserAttributes: userAttributes.concat([
            {
                Name: "custom:nocCodes",
                Value: nocCodes.toString(),
            },
        ]),
    };

    const createUserResult =
        userData.group === "operators"
            ? await cognito.send(new AdminCreateUserCommand(createOperatorUserCommand))
            : await cognito.send(new AdminCreateUserCommand(createUserCommand));

    await cognito.send(
        new AdminAddUserToGroupCommand({
            GroupName: userData.group,
            Username: createUserResult.User?.Username,
            UserPoolId: userPoolId,
        }),
    );
};

export const initiateResetPassword = async (email: string) => {
    logger.info("", {
        context: "data.cognito",
        message: "Initiating reset password flow for cognito user",
    });

    try {
        const params: ForgotPasswordCommandInput = {
            Username: email,
            ClientId: cognitoClientId,
            SecretHash: calculateSecretHash(email),
        };
        return cognito.send(new ForgotPasswordCommand(params));
    } catch (error) {
        if (error instanceof Error) {
            throw new Error(`Failed to initiate reset password flow: ${error.stack || ""}`);
        }

        throw error;
    }
};

export const resetUserPassword = async (key: string, newPassword: string, email: string) => {
    logger.info("", {
        context: "data.cognito",
        message: "Resetting password for cognito user",
    });

    try {
        const params: ConfirmForgotPasswordCommandInput = {
            ClientId: cognitoClientId,
            ConfirmationCode: key,
            Password: newPassword,
            Username: email,
            SecretHash: calculateSecretHash(email),
        };
        return cognito.send(new ConfirmForgotPasswordCommand(params));
    } catch (error) {
        if (error instanceof Error) {
            throw new Error(`Failed to reset password: ${error.stack || ""}`);
        }

        throw error;
    }
};
