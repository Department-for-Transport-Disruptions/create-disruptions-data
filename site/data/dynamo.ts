import { DynamoDBClient } from "@aws-sdk/client-dynamodb";
import { DynamoDBDocumentClient, QueryCommand } from "@aws-sdk/lib-dynamodb";
import { ptSituationElementSchema } from "@create-disruptions-data/shared-ts/siriTypes.zod";
import { notEmpty } from "../utils";
import logger from "../utils/logger";

const ddbDocClient = DynamoDBDocumentClient.from(new DynamoDBClient({ region: "eu-west-2" }));

const tableName = process.env.TABLE_NAME as string;

export const getDisruptionsDataFromDynamo = async () => {
    logger.info("Getting disruptions data from DynamoDB table...");

    const dbData = await ddbDocClient.send(
        new QueryCommand({
            TableName: tableName,
            KeyConditionExpression: "PK = :i",
            ExpressionAttributeValues: {
                ":i": "1",
            },
        }),
    );

    return dbData.Items?.map((item) => {
        const parsedItem = ptSituationElementSchema.safeParse(item);

        if (!parsedItem.success) {
            logger.error("Error parsing disruption from dynamo");
            logger.error(parsedItem.error.stack);
            return null;
        }

        return parsedItem.data;
    }).filter(notEmpty);
};
