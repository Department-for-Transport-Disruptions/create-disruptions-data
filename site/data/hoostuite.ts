import startCase from "lodash/startCase";
import { NextPageContext } from "next";
import { parseCookies } from "nookies";
import { getParameter, getParametersByPath, putParameter } from "./ssm";
import { COOKIES_ID_TOKEN, COOKIES_REFRESH_TOKEN, HOOTSUITE_URL } from "../constants";
import { hootsuiteMeSchema, hootsuiteTokenSchema, hootsuiteSocialProfilesSchema } from "../schemas/hootsuite.schema";
import { HootsuiteProfiles, SocialMediaAccountsSchema } from "../schemas/social-media-accounts.schema";

export const getHootsuiteToken = async (refreshToken: string, authToken: string) => {
    return await fetch(`${HOOTSUITE_URL}oauth2/token`, {
        method: "POST",
        body: new URLSearchParams({
            grant_type: "refresh_token",
            refresh_token: refreshToken ?? "",
        }),
        headers: {
            "Content-Type": "application/x-www-form-urlencoded",
            Authorization: authToken,
        },
    });
};

export const getHootsuiteData = async (
    ctx: NextPageContext,
    username: string,
    orgId: string,
): Promise<{ clientId: string; userData: SocialMediaAccountsSchema }> => {
    const cookies = parseCookies(ctx);
    let clientIdValue = "";
    let userData: SocialMediaAccountsSchema = [];
    try {
        const idToken = cookies[COOKIES_ID_TOKEN];
        const refreshToken = cookies[COOKIES_REFRESH_TOKEN];

        const [clientId, clientSecret, keys, tokensByOrganisation] = await Promise.all([
            getParameter(`/social/hootsuite/client_id`),
            getParameter(`/social/hootsuite/client_secret`),
            getParametersByPath(`/social/${orgId}/hootsuite`),
            getParametersByPath(`/social/${orgId}/hootsuite`),
        ]);

        if (!clientId || !clientSecret) {
            throw new Error("clientId and clientSecret must be defined");
        }

        clientIdValue = clientId.Parameter?.Value || "";
        const key = `${clientId.Parameter?.Value || ""}:${clientSecret.Parameter?.Value || ""}`;

        const authToken = `Basic ${Buffer.from(key).toString("base64")}`;

        if (idToken && refreshToken)
            await Promise.all([
                putParameter(`/${username}/token`, idToken, "SecureString", true),
                putParameter(`/${username}/refresh-token`, refreshToken, "SecureString", true),
            ]);

        const refreshTokens = tokensByOrganisation?.Parameters?.map((token) => {
            return {
                value: token.Value,
                name: token.Name,
                userId: token?.Name?.split("hootsuite/")[1]?.split("-")[0] ?? "",
                accountType: startCase(token?.Name?.split("/")[3]) ?? "",
                addedBy: token?.Name?.split("/")[4]?.replace("_", " ")?.split("-")[1] ?? "",
            };
        });

        if (refreshTokens && refreshTokens.length > 0) {
            await Promise.all(
                refreshTokens?.map(async (token) => {
                    const resp = await getHootsuiteToken(token.value || "", authToken);
                    if (resp.ok) {
                        const tokenResult = hootsuiteTokenSchema.parse(await resp.json());

                        if (!keys || (refreshTokens && keys.Parameters?.length === 0)) {
                            throw new Error("Refresh token is required to fetch dropdown data");
                        }
                        const key: string =
                            keys.Parameters?.find((rt) => rt.Name?.includes(`${token.userId}`))?.Name || "";
                        if (!key) {
                            throw new Error("Refresh token is required to fetch dropdown data");
                        }

                        await putParameter(key, tokenResult.refresh_token ?? "", "SecureString", true);
                        const userDetailsResponse = await fetch(`${HOOTSUITE_URL}v1/me`, {
                            method: "GET",
                            headers: {
                                Authorization: `Bearer ${tokenResult.access_token ?? ""}`,
                            },
                        });
                        if (userDetailsResponse.ok) {
                            const userDetails = hootsuiteMeSchema.parse(await userDetailsResponse.json());
                            const userInfo = userDetails.data || {};

                            const extraInfo = {
                                ...userInfo,
                                accountType: token.accountType || "",
                                addedBy: token.addedBy || "",
                                expiresIn: "Never",
                            };

                            const socialProfilesResponse = await fetch(`${HOOTSUITE_URL}v1/socialProfiles`, {
                                method: "GET",
                                headers: {
                                    Authorization: `Bearer ${tokenResult.access_token ?? ""}`,
                                },
                            });
                            if (socialProfilesResponse.ok) {
                                const socialProfiles = hootsuiteSocialProfilesSchema.parse(
                                    await socialProfilesResponse.json(),
                                );

                                userData = [
                                    ...userData,
                                    {
                                        ...extraInfo,
                                        hootsuiteProfiles:
                                            socialProfiles.data?.map((sp: HootsuiteProfiles[0]) => ({
                                                type: sp.type,
                                                socialNetworkId: sp.socialNetworkId,
                                                id: sp.id,
                                            })) ?? [],
                                    },
                                ];
                            }
                        }
                    }
                }),
            );
        }

        return { clientId: clientIdValue, userData };
    } catch (e) {
        throw new Error(`${(e as Error).message}`);
    }
};
