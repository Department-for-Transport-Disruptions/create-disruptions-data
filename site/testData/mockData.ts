/* eslint-disable @typescript-eslint/no-unsafe-assignment */
/* eslint-disable @typescript-eslint/no-explicit-any */
import { mockRequest, mockResponse } from "mock-req-res";
import { NextPageContext } from "next";
import React from "react";
import { Mock, vi } from "vitest";
import { COOKIE_ID_TOKEN, COOKIES_POLICY_COOKIE } from "../constants";

export interface GetMockContextInput {
    session?: { [key: string]: any };
    cookies?: any;
    body?: any;
    url?: any;
    uuid?: any;
    mockWriteHeadFn?: Mock<any, any>;
    mockEndFn?: Mock<any, any>;
    isLoggedin?: boolean;
    query?: any;
}

export const getMockRequestAndResponse = ({
    cookieValues = {},
    body = null,
    mockWriteHeadFn = vi.fn(),
    mockEndFn = vi.fn(),
    requestHeaders = {},
    isLoggedin = true,
    url = null,
    query = null,
}: GetMockRequestAndResponse = {}): { req: any; res: any } => {
    const res = mockResponse({ writeHead: mockWriteHeadFn, end: mockEndFn });

    const {
        idToken = "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJjdXN0b206bm9jIjoiVEVTVCIsImVtYWlsIjoidGVzdEBleGFtcGxlLmNvbSIsImp0aSI6Ijg1MmQ1MTVlLTU5YWUtNDllZi1iMTA5LTI4YTRhNzk3YWFkNSIsImlhdCI6MTU5Mjk4NzMwNywiZXhwIjoxNTkyOTkwOTA3fQ.DFdxnpdhykDONOMeZMNeMUFpCHZ-hQ3UXczq_Qh0IAI",
        cookiePolicy = null,
    } = cookieValues;

    const defaultSession = {
        // eslint-disable-next-line @typescript-eslint/no-empty-function
        destroy: (): void => {},
    };

    let cookieString = "";

    cookieString += isLoggedin ? `${COOKIE_ID_TOKEN}=${idToken as string};` : "";

    cookieString += cookiePolicy ? `${COOKIES_POLICY_COOKIE}=${encodeURI(JSON.stringify(cookiePolicy))}` : "";

    const req = mockRequest({
        connection: {
            encrypted: true,
        },
        url,
        headers: {
            host: "localhost:3000",
            cookie: cookieString,
            origin: "localhost:3000",
            ...requestHeaders,
        },
        cookies: cookieValues,
        session: { ...defaultSession },
    });

    if (body) {
        req.body = body;
    }
    if (query) {
        req.query = query;
    }

    return { req, res };
};

export const getMockContext = ({
    cookies = {},
    body = null,
    uuid = {},
    mockWriteHeadFn = vi.fn(),
    mockEndFn = vi.fn(),
    isLoggedin = true,
    url = null,
    query = "",
}: GetMockContextInput = {}): NextPageContext => {
    const { req, res } = getMockRequestAndResponse({
        cookieValues: cookies,
        body,
        uuid,
        mockWriteHeadFn,
        mockEndFn,
        requestHeaders: {},
        isLoggedin,
        url,
    });

    const ctx: NextPageContext = {
        res,
        req,
        pathname: "",
        query,
        // eslint-disable-next-line react/display-name
        AppTree: () => React.createElement("div"),
    };

    return ctx;
};

export interface GetMockRequestAndResponse {
    cookieValues?: any;
    body?: any;
    uuid?: any;
    mockWriteHeadFn?: Mock<any, any>;
    mockEndFn?: Mock<any, any>;
    requestHeaders?: any;
    isLoggedin?: boolean;
    url?: any;
    query?: any;
}
