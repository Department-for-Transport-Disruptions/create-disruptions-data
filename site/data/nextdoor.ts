import { getOrgSocialAccounts } from "./dynamo";
import { SocialMediaAccount } from "../schemas/social-media-accounts.schema";

export const getNextdoorAuthUrl = () => {
    return "";
};

export const getNextdoorAccountList = async (orgId: string, operatorOrgId?: string): Promise<SocialMediaAccount[]> => {
    const socialAccounts = await getOrgSocialAccounts(orgId);

    const nextdoorDetail = socialAccounts
        .filter((account) => account.accountType === "Nextdoor")
        .map((account) => ({
            ...account,
            expiresIn: "Never",
        }))
        .filter((item) =>
            operatorOrgId ? operatorOrgId === item.createdByOperatorOrgId : !item.createdByOperatorOrgId,
        );

    return nextdoorDetail;
};

export const getNextdoorGroupIds = (): number[] => [1, 2];
