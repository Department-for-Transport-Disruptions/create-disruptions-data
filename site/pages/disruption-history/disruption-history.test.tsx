import { History } from "@create-disruptions-data/shared-ts/disruptionTypes.zod";
import { PublishStatus } from "@create-disruptions-data/shared-ts/enums";
import MockDate from "mockdate";
import renderer from "react-test-renderer";
import { afterAll, describe, expect, it, vi } from "vitest";
import * as dynamo from "../../data/dynamo";
import { DEFAULT_DISRUPTION_ID, disruptionWithConsequences, getMockContext } from "../../testData/mockData";
import DisruptionHistory, { getServerSideProps } from "./[disruptionId].page";

const history: History[] = [
    {
        datetime: "2023-05-19T14:45:00Z",
        status: PublishStatus.published,
        user: "Test User 3",
        historyItems: ["Some Text Again", "Some More Text"],
    },
    {
        datetime: "2023-05-19T14:40:00Z",
        status: PublishStatus.pendingApproval,
        user: "Test User 1",
        historyItems: ["Some Text"],
    },
];

const historyToSort: History[] = [
    {
        datetime: "2023-05-19T14:40:00Z",
        status: PublishStatus.pendingApproval,
        user: "Test User 1",
        historyItems: ["Some Text"],
    },
    {
        datetime: "2023-05-11T14:40:00Z",
        status: PublishStatus.pendingApproval,
        user: "Test User 1",
        historyItems: ["Some Text"],
    },
    {
        datetime: "2023-05-13T14:45:00Z",
        status: PublishStatus.published,
        user: "Test User 3",
        historyItems: ["Some Text Again", "Some More Text"],
    },
];

describe("pages", () => {
    const getDisruptionSpy = vi.spyOn(dynamo, "getDisruptionById");

    vi.mock("../../data/dynamo", () => ({
        getDisruptionById: vi.fn(),
    }));

    MockDate.set("2023-05-22");

    afterAll(() => {
        MockDate.reset();
    });

    describe("DisruptionHistory", () => {
        it("should render correctly", () => {
            const tree = renderer
                .create(<DisruptionHistory disruptionId={DEFAULT_DISRUPTION_ID} history={history} />)
                .toJSON();
            expect(tree).toMatchSnapshot();
        });
    });

    describe("getServerSideProps", () => {
        it("should return correctly sorted history items", async () => {
            getDisruptionSpy.mockResolvedValue({
                ...disruptionWithConsequences,
                history: historyToSort,
            });
            const props = await getServerSideProps(getMockContext());

            expect(props).toEqual({
                props: {
                    disruptionId: disruptionWithConsequences.disruptionId,
                    history: [
                        {
                            datetime: "2023-05-19T14:40:00Z",
                            status: "PENDING_APPROVAL",
                            user: "Test User 1",
                            historyItems: ["Some Text"],
                        },
                        {
                            datetime: "2023-05-13T14:45:00Z",
                            status: "PUBLISHED",
                            user: "Test User 3",
                            historyItems: ["Some Text Again", "Some More Text"],
                        },
                        {
                            datetime: "2023-05-11T14:40:00Z",
                            status: "PENDING_APPROVAL",
                            user: "Test User 1",
                            historyItems: ["Some Text"],
                        },
                    ],
                },
            });
        });

        it("should add history item if disruption has closed", async () => {
            getDisruptionSpy.mockResolvedValue({
                ...disruptionWithConsequences,
                disruptionNoEndDateTime: "",
                disruptionEndDate: "20/05/2023",
                disruptionEndTime: "1200",
                publishEndDate: "20/05/2023",
                publishEndTime: "1200",
                history: historyToSort,
            });
            const props = await getServerSideProps(getMockContext());

            expect(props).toEqual({
                props: {
                    disruptionId: disruptionWithConsequences.disruptionId,
                    history: [
                        {
                            datetime: "2023-05-20T11:00:00.000Z",
                            historyItems: ["Changed status: Open to closed"],
                            status: "PUBLISHED",
                            user: "System",
                        },
                        {
                            datetime: "2023-05-19T14:40:00Z",
                            status: "PENDING_APPROVAL",
                            user: "Test User 1",
                            historyItems: ["Some Text"],
                        },
                        {
                            datetime: "2023-05-13T14:45:00Z",
                            status: "PUBLISHED",
                            user: "Test User 3",
                            historyItems: ["Some Text Again", "Some More Text"],
                        },
                        {
                            datetime: "2023-05-11T14:40:00Z",
                            status: "PENDING_APPROVAL",
                            user: "Test User 1",
                            historyItems: ["Some Text"],
                        },
                    ],
                },
            });
        });
    });
});
