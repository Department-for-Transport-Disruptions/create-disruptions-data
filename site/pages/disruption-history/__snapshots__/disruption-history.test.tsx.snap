// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`pages > DisruptionHistory > should render correctly 1`] = `
<DocumentFragment>
  <div
    className="govuk-phase-banner max-w-[960px] w-[90%] border-0 mx-auto app-phase-banner app-width-container"
  >
    <p
      className="govuk-phase-banner__content"
    >
      <strong
        className="govuk-tag govuk-phase-banner__content__tag"
      >
        Beta
      </strong>
      <span
        className="govuk-phase-banner__text"
      >
        This is a new service. Help us improve it and
         
        <a
          className="govuk-link"
          href="/feedback"
          id="feedback-link"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onTouchStart={[Function]}
        >
          give your feedback (opens in new tab)
        </a>
        .
      </span>
    </p>
  </div>,
  <div
    class="govuk-width-container"
  >
    <button
      class="govuk-back-link mb-7"
      type="button"
    >
      Back
    </button>
    <main
      class="govuk-main-wrapper"
      id="main-content"
    >
      <h1
        class="govuk-heading-xl"
      >
        Disruption history
      </h1>
      <table
        class="govuk-table"
      >
        <thead
          class="govuk-table__head"
        >
          <tr
            class="govuk-table__row"
          />
        </thead>
        <tbody
          class="govuk-table__body"
        >
          <tr
            class="govuk-table__row"
          >
            <td
              class="govuk-table__cell align-middle overflow-hidden px-2 max-w-2xl "
            >
              <h2
                class="govuk-heading-s mb-3"
              >
                Test User 3, 2 days ago
              </h2>
              <p
                class="govuk-body mb-0"
              >
                Some Text Again
              </p>
              <p
                class="govuk-body mb-0"
              >
                Some More Text
              </p>
              <p
                class="govuk-body mb-0"
              >
                Changed status: Draft pending approval to Open
              </p>
            </td>
            <td
              class="govuk-table__cell align-middle overflow-hidden px-2 max-w-2xl "
            >
              <p
                class="text-right"
              >
                19/05/2023 15:45
              </p>
            </td>
          </tr>
          <tr
            class="govuk-table__row"
          >
            <td
              class="govuk-table__cell align-middle overflow-hidden px-2 max-w-2xl "
            >
              <h2
                class="govuk-heading-s mb-3"
              >
                Test User 1, 2 days ago
              </h2>
              <p
                class="govuk-body mb-0"
              >
                Some Text
              </p>
              <p
                class="govuk-body mb-0"
              >
                Status: Draft pending approval
              </p>
            </td>
            <td
              class="govuk-table__cell align-middle overflow-hidden px-2 max-w-2xl "
            >
              <p
                class="text-right"
              >
                19/05/2023 15:40
              </p>
            </td>
          </tr>
        </tbody>
      </table>
      <a
        class="govuk-button mt-8 govuk-button govuk-button--secondary"
        href="/disruption-detail/8befe1e9-e317-45af-825a-e0254fabf49d"
        role="button"
      >
        Back to details
      </a>
    </main>
    <div>
      <h2
        class="govuk-heading-s"
      >
        Help and Support
      </h2>
      <p
        class="govuk-body"
      >
        If you are having problems, please contact the Create Disruption Service via this link: 
        <a
          class="govuk-link govuk-!-font-size-19"
          href="/contact"
        >
          Contact us
        </a>
      </p>
    </div>
  </div>
  <footer
    class="govuk-footer"
    role="contentinfo"
  >
    <div
      class="govuk-width-container"
    >
      <div
        class="govuk-footer__meta"
      >
        <div
          class="govuk-footer__meta-item govuk-footer__meta-item--grow"
        >
          <h2
            class="govuk-visually-hidden"
          >
            Support links
          </h2>
          <ul
            class="govuk-footer__inline-list"
          >
            <li
              class="govuk-footer__inline-list-item"
            >
              <a
                class="govuk-footer__link"
                href="/contact"
              >
                Help
              </a>
            </li>
            <li
              class="govuk-footer__inline-list-item"
            >
              <a
                class="govuk-footer__link"
                href="/cookies"
              >
                Cookies
              </a>
            </li>
            <li
              class="govuk-footer__inline-list-item"
            >
              <a
                class="govuk-footer__link"
                href="/contact"
              >
                Contact
              </a>
            </li>
            <li
              class="govuk-footer__inline-list-item"
            >
              <a
                class="govuk-footer__link"
                href="/accessibility"
              >
                Accessibility
              </a>
            </li>
            <li
              class="govuk-footer__inline-list-item"
            >
              <a
                class="govuk-footer__link"
                href="https://www.gov.uk/help/terms-conditions"
              >
                Terms and conditions
              </a>
            </li>
            <li
              class="govuk-footer__inline-list-item"
            >
              <a
                class="govuk-footer__link"
                href="/privacy"
              >
                Privacy
              </a>
            </li>
          </ul>
          <p
            class="govuk-footer__body"
          >
            Built by the 
            <a
              class="govuk-footer__link"
              href="https://www.gov.uk/government/organisations/department-for-transport"
              rel="noreferrer"
              target="_blank"
            >
              Department for Transport
            </a>
          </p>
          <svg
            aria-hidden="true"
            class="govuk-footer__licence-logo"
            focusable="false"
            height="17"
            viewBox="0 0 483.2 195.7"
            width="41"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M421.5 142.8V.1l-50.7 32.3v161.1h112.4v-50.7zm-122.3-9.6A47.12 47.12 0 0 1 221 97.8c0-26 21.1-47.1 47.1-47.1 16.7 0 31.4 8.7 39.7 21.8l42.7-27.2A97.63 97.63 0 0 0 268.1 0c-36.5 0-68.3 20.1-85.1 49.7A98 98 0 0 0 97.8 0C43.9 0 0 43.9 0 97.8s43.9 97.8 97.8 97.8c36.5 0 68.3-20.1 85.1-49.7a97.76 97.76 0 0 0 149.6 25.4l19.4 22.2h3v-87.8h-80l24.3 27.5zM97.8 145c-26 0-47.1-21.1-47.1-47.1s21.1-47.1 47.1-47.1 47.2 21 47.2 47S123.8 145 97.8 145"
              fill="currentColor"
            />
          </svg>
          <span
            class="govuk-footer__licence-description"
          >
            All content is available under the 
            <a
              class="govuk-footer__link"
              href="https://www.nationalarchives.gov.uk/doc/open-government-licence/version/3/"
              rel="license"
            >
              Open Government Licence v3.0
            </a>
            , except where otherwise stated
          </span>
        </div>
        <div
          class="govuk-footer__meta-item"
        >
          <a
            class="govuk-footer__link govuk-footer__copyright-logo bg-govuk-crest"
            href="https://www.nationalarchives.gov.uk/information-management/re-using-public-sector-information/uk-government-licensing-framework/crown-copyright/"
          >
            Â© Crown copyright
          </a>
        </div>
      </div>
    </div>
  </footer>
</DocumentFragment>
`;
